test/gateway/eip20_cogateway/progress_revert_redeem.js

```diff
@@ -0,0 +1,385 @@
+    );
+
+    assert.strictEqual(
+      redeemerFinalTokenBalance.eq(redeemerInitialTokenBalance.add(revertRedeemParams.amount)),
+      true,
+      `Redeemer balance must increase by ${revertRedeemParams.amount.toString(10)}.`,
+    );
+
+    assert.strictEqual(
+      burnerFinalTokenBalance.eq(burnerInitialTokenBalance),
+      true,
+      `Burner balance must not change.`,
+    );
+
+  });
+
+  it('should fail when message hash is zero', async function () {
+
+    await cogateway.setStorageRoot(revertRedeemParams.blockHeight, revertRedeemParams.storageRoot);
+
+    await Utils.expectRevert(
+      cogateway.progressRevertRedeem(
+        ZeroBytes,
+        revertRedeemParams.blockHeight,
+        revertRedeemParams.rlpParentNodes,
+      ),
+      'Message hash must not be zero.',
+    );
+
+  });
+
+  it('should fail when storage proof zero', async function () {
+
+    await cogateway.setStorageRoot(revertRedeemParams.blockHeight, revertRedeemParams.storageRoot);
+
+    await Utils.expectRevert(
+      cogateway.progressRevertRedeem(
+        revertRedeemParams.messageHash,
+        revertRedeemParams.blockHeight,
+        '0x',
+      ),
+      'RLP parent nodes must not be zero.',
+    );
+  });
+
+  it('should fail when storage proof is invalid', async function () {
+
+    await cogateway.setStorageRoot(revertRedeemParams.blockHeight, revertRedeemParams.storageRoot);
+
+    await Utils.expectRevert(
+      cogateway.progressRevertRedeem(
+        revertRedeemParams.messageHash,
+        revertRedeemParams.blockHeight,
+        "0x1245",
+      ),
+      'VM Exception while processing transaction: revert',
+    );
+
+  });
+
+  it('should fail when storage root is not committed for given block height', async function () {
+
+    await Utils.expectRevert(
+      cogateway.progressRevertRedeem(
+        revertRedeemParams.messageHash,
+        revertRedeemParams.blockHeight,
+        revertRedeemParams.rlpParentNodes,
+      ),
+      'Storage root must not be zero.',
+    );
+
+  });
+
+  it('should fail when message is undeclared', async function () {
+
+    await Utils.expectRevert(
+      cogateway.progressRevertRedeem(
+        web3.utils.sha3("dummy"),
+        revertRedeemParams.blockHeight,
+        revertRedeemParams.rlpParentNodes,
+      ),
+      'RedeemIntentHash must not be zero.',
+    );
+
+  });
+
+  it('should fail when message is progressed', async function () {
+
+    await cogateway.setOutboxStatus(
+      revertRedeemParams.messageHash,
+      MessageStatusEnum.Progressed,
+    );
+
+    await cogateway.setStorageRoot(revertRedeemParams.blockHeight, revertRedeemParams.storageRoot);
+
+    await Utils.expectRevert(
+      cogateway.progressRevertRedeem(
+        revertRedeemParams.messageHash,
+        revertRedeemParams.blockHeight,
+        revertRedeemParams.rlpParentNodes,
+      ),
+      'Message on source must be DeclaredRevocation.',
```

Comment:
Should probably `must have status` instead of `must be` 😳

Reply:
How about `Message status on source must be DeclaredRevocation.` ?

Reply:
👍